### DISCUSSION TITLE
Dynamic layout generation

### DISCUSSION THREAD

**Discussion by sylzys** [orchidsoftware/platform#1843] [Q&A] [answered]
Is there a way to generate layouts dynamycally ?

I need to generate X layouts based on amount of days in months. I tried do do
``` 
$layout = Layout::tabs([]);
$layout['tab1'] = MyCustomLayout::class;

``` 

But got `Cannot use object of type class@anonymous as array `.

I managed to generate everything inside a custom blade template with query object, loops and all, but my form can't send a POST request, got an error saying `method not supported`.

Absolutly no idea what to try :/

@tabuna, any ideas to help please ? Thanks in advance

**Comment by tabuna** [comment, marked as answer]
Hi @sylzys, classes for layers are not intended for construction, but you can move the logic to any other place and pass a list of classes to display, like this:

```php
public function layout(): array
{
    $tabs = collect([
        Tab1::class,
        Tab2::class,
        Tab3::class,
    ])
        ->filter(function (string $class) {
            return $class !== Tab1::class;
        })
        ->toArray();
    
    return [
        Layout::tabs($tabs)
    ];
}
```


**A slightly more complex example of construction, not for your case. But I must mention:**


Personally, I try to avoid this definition, as I want to open the screen and see all of its definitions, not conditions. 
For this, I prefer to define the mapping directly in the classes, for example:

```php
php artisan orchid:rows ConditionShowExample
```

And in it, define the `isSee` method

```php
<?php

namespace App\Orchid\Layouts;

use Orchid\Screen\Field;
use Orchid\Screen\Layouts\Rows;

class ConditionShowExample extends Rows
{
    /**
     * Used to create the title of a group of form elements.
     *
     * @var string|null
     */
    protected $title;

    /**
     * Get the fields elements to be displayed.
     *
     * @return Field[]
     */
    protected function fields(): array
    {
        return [];
    }

    /**
     * @return bool
     */
    public function isSee(): bool
    {
        // Any conditions for display
        // Screen data can be accessed via $this->query
       // https://orchid.software/en/recipes/how-to-access-screen-data-from-a-layer/

        return true;
    }
}
```

Ultimately, this allows you to keep the screen clean and easy to read, leaving different definitions:


```php
public function layout(): array
{
    return [
        Layout::tabs([
            ConditionShowExample::class, // Each of them can have its own `isSee` condition
            ConditionShowExample::class, // Each of them can have its own `isSee` condition
            ConditionShowExample::class, // Each of them can have its own `isSee` condition
        ]),
    ];
}
```

It doesn't matter which method you choose, but I hope my answer was helpful and suggested how to use it.

**Reply by sylzys** [reply]
Thanks a lot for this explanation :)

The difficulty I faced was that I need  to deal with imbricated tabs, and not always the same length.

e.g, August 2021 has 6 weeks,  first 1 with only 1 day, 4 weeks with 7 days and last week with 3 days, so I need to come up with a layout like
| Week 1 | Week 2 | Week 3 | Week 4 | Week 5 |
|--------|--------|--------|--------|--------|

then on Week 1:

| Week 1 |       |       |       |    |
|--------|-------|-------|-------|----|
| **Day 1**  |       |       |       |    |

Week 2:

|  |    Week 2   |       |       |    |
|--------|-------|-------|-------|----|
| **Day 2**  |    Day 3   |  Day 4     |       |    |

and so on.. with fields() inside each day.
Can I achieve this with collect() ? 

I understand I could use 1 collect for weeks then 1 collect() in each for days I guess, but how could I deal with amount of days in each week to get the exact number of tabs, depending on each month ? 

I hope my question is clear, thanks again for your help @tabuna

**Reply by sylzys** [reply]
hello @tabuna , when I try your first example I get the following error : 
`argument 1 passed to Orchid\Screen\LayoutFactory::tabs() must be of the type array, object given, called in D:\slab\vendor\laravel\framework\src\Illuminate\Support\Facades\Facade.php on line 261 `

I really don't know how to do here.. As I told you  I managed to recreate everythin in the layout using PHP + JS but can't submit the form.. Is there anything I should specify on my screen to force a POST method ? (the error is `method not supported`

Thanks again
